### DISCUSSION TITLE
transforming `<p>honk<br /></p>` into Markdown yields a stray backslash?

### DISCUSSION THREAD

**Discussion by cpiro** [syntax-tree/unist#50] [Q&A] [answered]
Hi all! I've been using these tools to turn the HTML output from a [tiptap](https://www.tiptap.dev/)-based editor into Markdown. In particular, I've been seeing weird behavior when the editor document ends in a line break: 

1. The editor outputs `<p>honk<br /></p>`. This seems to be a valid HTML fragment.
2. `rehypeParse` parses that in the expected way, giving a hast with a `p` tag containing the text then a `br` tag.
3. `rehype2remark` / `hast-util-to-mdast` transforms that hast into the mdast with a `paragraph` containing the text then a `break`.
4. `remarkStringify` transforms that mdast into the Javascript string `"honk\\\n"`, that is, four characters followed by a single backslash then a LF.

The commonmark.org dingus says that the output of 4. is the Markdown that means the text "honk\\" (five characters, the last of which is a backslash). `remarkParse` agrees that the output of 4. corresponds to the a `paragraph` containing only the five characters of text "honk\\", and indeed the [CommonMark spec 6.9](https://spec.commonmark.org/0.17/#hard-line-breaks) says that

> Hard line breaks are for separating inline content within a block. Neither syntax for hard line breaks works at the end of a paragraph or other block element

Clearly I don't want the Markdown output to have a stray backslash character at the end when the editor has no backslashes in it. The problem seems to be that it's OK to have a paragraph that ends in a line break in HTML, but in CommonMark it is not. In 3. `hast-util-to-mdast` transforms the ASTs in the straightforward way, but is the output a "valid" mdast? Or is it the responsibility in 4. for `remarkStringify` not to emit a stray backslash for its valid input? Of course I could sanitize the HTML on the way out of the editor to remove trailing `br`s from a `p`, but this seems to be valid HTML. Or is there something I'm missing altogether?

Thank you! 

Here's the code I used to test:

```js
const html = "<p>honk<br /></p>";
const hast = unified().use(rehypeParse).parse(html);
const mdast = toMdast(hast); // from "hast-util-to-mdast"
const md = unified().use(remarkStringify).stringify(mdast);

const check = unified().use(remarkParse).parse(md);
```

There's a [codesandbox.io doc](https://codesandbox.io/s/holy-sea-hg7xc) with the above, and here's a screenshot of the console output (with some of the AST fields stripped out for readability):
![screenshot](https://user-images.githubusercontent.com/91721/128572173-382e9b61-662e-4fa8-9d1b-76f87c549909.png)

**Comment by wooorm** [comment, marked as answer]
Hi there 👋

Yeah, it’s valid HTML. I think this would be something for `hast-util-to-mdast`.
The same could be done with headings: `<h1>x<br></h1>`

I think it would be better to just remove them, instead of the alternative of injecting some non-breaking invisible character after them.

Anyway, probably a `filter-trailing-breaks` or so utility in hast-util-to-mdast used for both headings andf paragraphs?
